### ISSUE TITLE
Random_Suffle causes hanging at a specific batch step

### ISSUE THREAD

**Issue by psandev** [NVIDIA/DALI#3348]
Hi guys,
the dataloader hangs always exactly at a specific step when I use random_shuffle=True. I played with initial_fill (11-50) and num_threads (2-6, I have 16 threaded CPU), but nothing helps. 
Is this a bug or is it a matter of configuration parameters? 
I am using the lates 1.5.0 version.

**Comment by psandev** [comment]
Also, I wanted to downgrade to 1.4.0 to check if the issue exists there as well, but I couldn't find anywhere in the docs how to install a specific version. Could you send me a link? Thanks a lot.

**Comment by JanuszL** [comment]
Hi @psandev,

> I wanted to downgrade to 1.4.0 to check if the issue exists there as well

You can do as with any other pip package, just provide the version you want to install explicitly:
```
pip install --extra-index-url https://developer.download.nvidia.com/compute/redist --upgrade nvidia-dali-cuda110==1.4.0
```

> the dataloader hangs always exactly at a specific step when I use random_shuffle=True. I played with initial_fill (11-50) and

Can you provide more details and answer the below questions:
- What dataset and data loader do you use?
- Does it happen when you turn off the shuffling?
- Have you checked how CPU and IO (disc access) activity looks like? Maybe part of your data set is cached in the disc cache while the access to the other part of the data requires read which could be terribly slow and that can make an impression of hand.
- Where your data is located? Local storage or remote one?

**Comment by psandev** [comment]
Hi,
it took me one and a half day, but I got to the bottom of it. The issue is not related to the random_shuffle. It happens with random_shuffle=False as well, but towards the end of the epoch, which is ~ 6 hours, so when I created the issue, I did not reach the point at which the dataloader hang. i just did not have the patience. Sorry about that.
The issue is with my data and I am really happy that this happened, because it [pointed out to some other issues, which I couldn't understand. 
My data is video clips of length 16 and more frames each. The problem happens on clips that are shorter than 16 frames, which were not supposed to be there. I was unaware of their existence.
I use fn.readers.video_resize with sequence_length of 8, stride of 2 and step of 4. This should create 3 clips of 8 frames each, with the last frame of the last clip being the 16th frame of the original clip. So in a clip of less tha 16 frames, the third sub-clip cannot be created due to the missing frames. 
The thing is that I would expect the dataloader to fail with appropriate failure message. maybe you can add a failure mode.
I am closing the issue. Let me know if I can assist in case you want to work on it. 
Thanks a gain for the quick response, as usual.

**Comment by JanuszL** [comment]
Hi @psandev,

> I use fn.readers.video_resize with sequence_length of 8, stride of 2 and step of 4. On a 16 frames long clip, this should create 3 clips of 8 frames each, with the last frame of the last clip being the 16th frame of the original clip. So if a clip of less than 16 frames long, the third sub-clip cannot be created due to the missing frames.

If I do the math correctly, in the case of 16 frame video you should get only the sequences [0, 2, 4, 6,8, 10, 12, 14], the next one would start from frame 4 and cannot be created unless you set `pad_sequences=True`.
If the sequence cannot be created the loader would just skip the file. If you see a hang this must be something else.

Have you rerun the test using only the file you suspect that causes this? Can you share it with us so we can retest it on our side?

**Comment by psandev** [comment]
My bad. The stride was 1. Sorry. I did many tests, so I got a bit mixed up. I deleted already the defective clips, but you can create some clips with 16 and 15 frames length. I can send you some of mine, if you need. I did not play with the padding. 
I use pytorch. DALIClassificationIterator is used with its default parameters.

**Comment by JanuszL** [comment]
Hi @psandev,

I just checked with the DALI_extra video: db/video/sintel/video_files/sintel_trailer-720p_0.mp4. When I asked to create a sequence of length 242 (equal to its total length), the `pipeline.epoch_size()` returns 1, while when asked for 243 I receive expected error `There are no valid sequences in the provided dataset, check the length of the available videos and the requested sequence length.`. 
From my point of view, it works as expected.

**Comment by psandev** [comment]
Thank you for the check. On the other hand, the problem disappeared once I cleaned the dataset. I will try to reproduce it by adding some shorter clips.

**Comment by JanuszL** [comment]
It could be just a video sample that is not well handled by DALI (like variable frame rate one).

**Comment by psandev** [comment]
I dont't have those. I cut the clips from big movies using OpenCV and saved them with cv2.VideoWriter at 25fps.
